blocks/graph.py

```diff
@@ -37,6 +37,7 @@ def _get_variables(self):
         self.applies = set()
         self.application_calls = set()
         self.updates = []
+        self.auxiliary_variables = set()
```

Comment:
Okay, I see, I had a rather different plan for L2 and L1: have a routine `add_L2_regularization(cost, params)` where cost is the cost Theano variable, and `params` are the parameters that you want to include in the penalty calculation. This routine would return a new Theano variable corresponding to the cost with regularization. 

So when I want to apply this regularization only for parameters of certain bricks I first use `Selector` to choose those bricks and extract their parameters and then pass this parameters to `add_L2_regularization`.

I like this plan more than what you propose, because when separation in a few consecutive steps possible, it should be done. When it is not possible, that is when the regularization term only makes sense in the context of a particular brick, the brick can add this term to the graph as an auxiliary variable.

I can do a pull request for L2 if you want.